### PR TITLE
[BEAM-10925] Load SQL UDFs from jar.

### PR DESCRIPTION
**Pull request by ibzib**
**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by ibzib** [comment]
R: @amaliujia @apilloud 

There should be ~1 more PR after this to finish the implementation of Java scalar UDFs (minus calc splitting).

**Comment by amaliujia** [comment]
Thanks! This is a relatively large PR so will spend more time to review tomorrow.

**Comment by apilloud** [comment]
I'm about to go on vacation, but I'll try to give this a quick pass tonight. It looks like nothing in here is ZetaSQL specific? If that is the case, you should move this to core SQL. Otherwise, how is this going to work for Calcite?

**Comment by ibzib** [comment]
> It looks like nothing in here is ZetaSQL specific? If that is the case, you should move this to core SQL.

Done.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+      }
+      return outputFile;
+    }
+  }
+
+  private ClassLoader createAndSetClassLoader(String inputJarPath) throws IOException {
+    File tmpJar = downloadFile(inputJarPath, "application/java-archive");
+    // Set the thread's context class loader so that the jar can be staged by the runner.
+    Thread.currentThread()
+        .setContextClassLoader(
```
Comment (apilloud) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java:126]: You've leaked the class loader here. setContextClassLoader should always be inside a try finally block that resets the class loader. For example: https://github.com/apache/beam/blob/3bb232fb098700de408f574585dfe74bbaff7230/sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamEnumerableConverter.java#L118

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+public class JavaUdfLoader {
+  private static final Logger LOG = LoggerFactory.getLogger(JavaUdfLoader.class);
+
+  /**
+   * Maps the external jar location to the functions the jar defines. Static so it can persist
+   * across multiple SQL transforms.
+   */
+  private static final Map<String, FunctionDefinitions> cache = new HashMap<>();
+
+  private static final ClassLoader originalClassLoader = ReflectHelpers.findClassLoader();
```
Comment (apilloud) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java:63]: This is probably unsafe to cache.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+ * <p>All UDFs are loaded and cached for each jar to mitigate IO costs.
+ */
+public class JavaUdfLoader {
+  private static final Logger LOG = LoggerFactory.getLogger(JavaUdfLoader.class);
+
+  /**
+   * Maps the external jar location to the functions the jar defines. Static so it can persist
+   * across multiple SQL transforms.
+   */
+  private static final Map<String, FunctionDefinitions> cache = new HashMap<>();
```
Comment (apilloud) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java:61]: I'm not sure this is safe as is. It will probably work for most cases, but I think you might want to key it on original class loader?

**Review by apilloud — APPROVED** [review]
LGTM minus the few class loader issues. Only the try-finally block is something that must be fixed before this goes in.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+  private static final ClassLoader originalClassLoader = ReflectHelpers.findClassLoader();
+
+  /**
+   * Load a user-defined scalar function from the specified jar.
+   *
+   * <p><strong>WARNING</strong>: The first time a jar is loaded, it is added to the thread's
+   * context {@link ClassLoader} so that the jar can be staged by the runner.
+   */
+  public ScalarFn loadScalarFunction(List<String> functionPath, String jarPath) {
```
Comment (amaliujia) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java:71]: I am a bit confused on what is `List<String> functionPath` (i.e. why it is a list?) Can you add a `@param` on java doc to explain this parameter?

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+
+  /**
+   * Creates a temporary local copy of the file at {@code inputPath}, and returns a handle to the
+   * local copy.
+   */
+  private File downloadFile(String inputPath, String mimeType) throws IOException {
+    Preconditions.checkArgument(!inputPath.isEmpty(), "Path cannot be empty.");
+    ResourceId inputResource = FileSystems.matchNewResource(inputPath, false /* is directory */);
+    try (ReadableByteChannel inputChannel = FileSystems.open(inputResource)) {
+      File outputFile = File.createTempFile("sql-udf-", inputResource.getFilename());
```
Comment (amaliujia) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java:104]: Will it better to add a randomization on temple file name?

I am thinking a case that there are two create function statement that each one specify a jar which has the same path but different buckets, so there might be a same local file name in this case?

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java (hunk 1 above)
Comment (ibzib) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java:126]: I intended to leave UDF provider jars on the class loader here, so that later, in the DF SQL launcher, we can use the class loader to detect which files need to be staged to the runner. https://github.com/apache/beam/blob/ecedd3e654352f1b51ab2caae0fd4665403bd0eb/runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineWorkerPoolOptions.java#L200

If this way is unsafe for some reason, we will have to come up with a different way to pass jars from JavaUdfLoader into pipeline options. The only other ways I could think of are:
1. Passing jars all the way up through the stack to SqlTransform, which will require changing SqlTransform and every other API surface in between.
2. Handling jars (independently from SqlTransform construction) a second time in the SQL launcher. In other words, parse the query text again to extract the UDF jars, download the UDF jars a second time, and add them to filesToStage.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java (hunk 2 above)
Comment (ibzib) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java:63]: Why?

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java (hunk 3 above)
Comment (ibzib) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java:61]: Why?

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java (hunk 5 above)
Comment (ibzib) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java:104]: `File.createTempFile` already produces a randomized name.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java (hunk 4 above)
Comment (ibzib) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java:71]: It's a list because functions in ZetaSQL can be period-separated (e.g. `foo.bar.baz` would be represented as a list of {"foo", "bar", "baz"}). I could change this to take a single String instead, I have no strong preference either way.

Comment (amaliujia) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java:71]: I see. List is ok as it is what ZetaSQL allows.

**Review by amaliujia — APPROVED** [review]
LGTM

After reading through this PR, as I recall there was a bunch of comments related UDF jar loading in the original prototype and those were addressed in https://github.com/apache/beam/pull/13200 already. So I don't have more comments really.

After existing open comments are addressed, this PR will be good to go.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java (hunk 1 above)
Comment (apilloud) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java:126]: Not resetting the ContextClassLoader after your method returns is unsafe. You need to find another way to communicate the classes upstream. Our current UDF loading happens in SqlTransform.expand(), seems like this should be called from there too. So option 1 should be trivial.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java (hunk 2 above)
Comment (apilloud) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java:63]: Users perform their own layering of class loaders. I've seen a tool that loads "stable" jars, such as released Beam Jars, into a root class loader that is reused across multiple pipelines, then loads the user code into a child class loader that is replaced for every pipeline. This could cause unexpected behavior as you'll probably end up caching the first child class loader to invoke this code.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java (hunk 3 above)
Comment (apilloud) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java:61]: See my comment about users creating their own class loaders above.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java (hunk 1 above)
Comment (ibzib) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java:126]: Sounds good. I will look into option 1 and probably implement it in a separate PR.

Comment (ibzib) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java:126]: Option 1 would require several API changes:
1. Add UDF provider jars to `BeamRelNode`.
2. Either change the signature of `ZetaSQLPlannerImpl::rel` (currently returns Calcite `RelRoot`) or give `ZetaSQLPlannerImpl` a mutable field to hold UDF provider jars.
3. Give SqlTransform a mutable field to hold UDF provider jars.

Of these, I'm most concerned with 3. Should PTransforms be mutable? I couldn't find anything in the Javadoc that explicitly warned against it, but mutability seems like something we should avoid if we can help it.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java (hunk 2 above)
Comment (ibzib) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java:63]: Got it. If we reset the class loader as discussed in the other thread, there will be no need for this cache.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java (hunk 1 above)
Comment (ibzib) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java:126]: A third option: if we download jars to a specific, known local directory, we could just add that entire directory to `filesToStage`.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+              "Failed to load user-defined scalar function %s from %s", functionFullName, jarPath),
+          e);
+    }
+  }
+
+  /**
+   * Creates a temporary local copy of the file at {@code inputPath}, and returns a handle to the
+   * local copy.
+   */
+  private File downloadFile(String inputPath, String mimeType) throws IOException {
```
Comment (apilloud) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java:100]: This method duplicates functionality built into the JVM. It might be possible to just construct a URL directly? If not,  here is an example that grabs the path to the temp file from the builtin URL jar cache:
`((JarURLConnection)new URL("jar:https://repo1.maven.org/maven2/com/google/api/gax-grpc/1.60.0/gax-grpc-1.60.0-guavashaded.jar!/").openConnection()).getJarFile().getName()`

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java (hunk 1 above)
Comment (apilloud) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java:126]: You've lost all isolation of the JARs on the worker, filesToStage is in the default classpath. I think you need to plumb the jar path all the way down to BeamCalcRel, and that is the layer at which you'll need to stage the jar and load it on the worker.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java (hunk 6 above)
Comment (ibzib) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java:100]: Thanks for the pointer. I'm not sure how GCS permissions work in this case, but it's worth a try.

Comment (apilloud) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java:100]: You'll have to use the GCS client to make an authenticated read off of GCS, the current code might be useful for that.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java (hunk 1 above)
Comment (amaliujia) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java:126]: I think the third option is the one that easy to implement and without obvious harms. 

I think the third option says to specify a known directory on the launcher VM (so not on workers).

Comment (ibzib) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java:126]: All of the options I listed were to address the question: how should we pass jars from JavaUdfLoader to the filesToStage pipeline option. Andrew's concern is bigger and applies to all three options: putting all the UDF jars in filesToStage puts them all on the classpath for workers. UDF jars are only needed when compiling the generated code in CalcFn, so they should not be on the classpath except for that.

Comment (amaliujia) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java:126]: As I recall we have discussed or at least mentioned such topic offline. 

For Java SDK, user's UDF is put into the classpath without an isolation. It won't be a regression for SQL if we follow up the same rule.

We could do better for the SQL case as we know the UDF will be executed in CalcFn, but do we need to handle it at this moment? Is it a required thing to launch the first version of UDF? IMO because SQL will be aligned with Java's practice thus it is not required.

Comment (ibzib) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java:126]: > For Java SDK, user's UDF is put into the classpath without an isolation. It won't be a regression for SQL if we follow up the same rule.

One difference is that in the Java SDK, the user is given direct control of the classpath via filesToStage. But when we construct a Java pipeline for them, the user does not have direct control of the classpath. The API we show to users does not expose how their Java transforms are fused into a single pipeline, nor should it. So isolating the classpath for each CalcFn better matches the API and hides complexity from the user.

> A further thought on jar isolation, will solution for the isolation in CalcFn for user-defined scalar func too specific. Will it better to have a separate discussion to cover also UDAF and UDTVF thus we can achieve the isolation on all types of UDF by a same solution?

That's a good question. Isolating the CalcFn is a nice solution for scalar functions, but I'm not sure how we could isolate UDAF since `CombineFn` is more abstract than `DoFn` and aggregates aren't compiled separately like CalcFn.

**Comment by ibzib** [comment]
Temporarily closing this as I am making some major changes.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java (hunk 6 above)
Comment (ibzib) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java:100]: @apilloud In order to use JarURLConnection we would have to write some code to transform a String (which is a Beam filesystem resource spec) into a jar URL. We could make the URL could point to either a local copy of the jar, or the original jar, which could be either local or remote.
- If we make and point to a local copy, the code wouldn't look too much different than it does currently.
- If we want to point to the original jar, our code would need to know how to convert a resource from any Beam filesystem into a URL. While that's possible, it would basically require adding a `getUrl` method to `ResourceId` and implementing it for every implementation of `ResourceId`.
 
tl;dr I'm not sure the benefit of using `JarURLConnection` outweighs the amount of additional code we'd need to write and maintain, since the current implementation supports every Beam filesystem for free.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java (hunk 3 above)
Comment (ibzib) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JavaUdfLoader.java:60]: I still can't imagine exactly how the user might break this by setting class loaders. I think we might just have to accept some level of risk here, since messing with the context class loader seems to be inherently kind of dangerous anyway.

**Comment by ibzib** [comment]
@apilloud @amaliujia I changed this to no longer set the context class loader as we discussed. If I understand correctly, the only place we really need the class loader to be set properly is in `CalcFn#compile`, so we can deal with that in a future PR. PTAL

**Comment by ibzib** [comment]
Run Java PreCommit

**Comment by ibzib** [comment]
PythonLint failure is obviously a flake, since there are 0 python changes: BEAM-11540

**Comment by amaliujia** [comment]
Thanks!  

+1 to not set the classloader and address when to set the classloader in another PR.

**Review by amaliujia — APPROVED** [review]

**Comment by suztomo** [comment]
"Run SQL Postcommit" fails in my PR https://github.com/apache/beam/pull/13804. @ibzib Do you think this PR is related?

```
java.lang.AssertionError: System property beam.sql.udf.test.jar_path must be set to run JavaUdfLoaderTest.
	at org.junit.Assert.fail(Assert.java:89)
	at org.apache.beam.sdk.extensions.sql.impl.JavaUdfLoaderTest.setUp(JavaUdfLoaderTest.java:50)
```

https://ci-beam.apache.org/job/beam_PostCommit_SQL_PR/lastBuild/testReport/junit/org.apache.beam.sdk.extensions.sql.impl/JavaUdfLoaderTest/testLoadScalarFunction_4/

**Comment by ibzib** [comment]
> "Run SQL Postcommit" fails in my PR #13804. @ibzib Do you think this PR is related?

It's definitely related (that test was newly added by this PR). I will look into it.

**Comment by suztomo** [comment]
Thank you.
